File: mobile/src/main/java/com/bclymer/dailybudget/fragments/BudgetsFragment.java
package com.bclymer.dailybudget.fragments;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.GridView;

import com.bclymer.dailybudget.R;
import com.bclymer.dailybudget.events.BudgetUpdatedEvent;
import com.bclymer.dailybudget.models.Budget;
import com.bclymer.dailybudget.utilities.ThreadManager;
import com.bclymer.dailybudget.views.BudgetView;
import com.nhaarman.listviewanimations.appearance.AnimationAdapter;
import com.nhaarman.listviewanimations.appearance.simple.AlphaInAnimationAdapter;
import com.nhaarman.listviewanimations.appearance.simple.ScaleInAnimationAdapter;
import com.nhaarman.listviewanimations.util.Insertable;

import java.util.ArrayList;
import java.util.List;

import butterknife.InjectView;
import butterknife.OnClick;
import butterknife.OnItemClick;

import static android.view.View.OnClickListener;
import static com.bclymer.dailybudget.fragments.EditBudgetFragment.NO_BUDGET_ID_VALUE;

/**
 * Created by bclymer on 9/26/2014.
 */
public class BudgetsFragment extends BaseFragment {

    @InjectView(android.R.id.list)
    protected GridView mGridView;
    @InjectView(android.R.id.empty)
    protected ViewGroup mEmptyView;

    private AnimationAdapter mAdapter;
    private List<Budget> mBudgetList;

    private BudgetSelectedCallback mCallback;

    public static BudgetsFragment newInstance() {
        return new BudgetsFragment();
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        if (!(activity instanceof BudgetSelectedCallback)) {
            throw new RuntimeException("Activity " + activity + " must implement BudgetSelectedCallback to display BudgetsFragment");
        } else {
            mCallback = (BudgetSelectedCallback) activity;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLayoutId = R.layout.fragment_budgets;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        // TODO: Loading indicator. Use Loader to load all budgets from DB. Maybe RxJava?
        mBudgetList = Budget.getDao().queryForAll();
        for (Budget budget : mBudgetList) {
            budget.updateCache();
        }

        BudgetAdapter budgetAdapter = new BudgetAdapter();
        mAdapter = new ScaleInAnimationAdapter(budgetAdapter);
        mAdapter.setAbsListView(mGridView);
        mGridView.setAdapter(mAdapter);
        mGridView.setEmptyView(mEmptyView);
        mEventBus.register(this);
    }

    @Override
    public void onDestroyView() {
        mEventBus.unregister(this);
        super.onDestroyView();
    }

    @OnItemClick(android.R.id.list)
    protected void onBudgetClick(int position) {
        mCallback.onBudgetSelected(mBudgetList.get(position).id);
    }

    @OnClick(R.id.fragment_budgets_emptyview_button_new_budget)
    protected void onCreateBudgetClick() {
        mCallback.onBudgetSelected(NO_BUDGET_ID_VALUE);
    }

    public void onEventMainThread(BudgetUpdatedEvent event) {
        List<Budget> copiedList = new ArrayList<>(mBudgetList);
        for (Budget budget : copiedList) {
            if (budget.id == event.budget.id) {
                if (event.deleted) {
                    mBudgetList.remove(event.budget);
                } else {
                    event.budget.cloneInto(budget);
                }
                mAdapter.notifyDataSetChanged();
                return;
            }
        }

        mBudgetList.add(event.budget);
        mAdapter.notifyDataSetChanged();
    }

    public interface BudgetSelectedCallback {
        public void onBudgetSelected(int budgetId);
    }

    private class BudgetAdapter extends BaseAdapter {

        private final LayoutInflater mInflater;

        public BudgetAdapter() {
            mInflater = getActivity().getLayoutInflater();
        }

        @Override
        public int getCount() {
            return mBudgetList != null ? mBudgetList.size() : 0;
        }

        @Override
        public Budget getItem(int position) {
            return mBudgetList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View view, ViewGroup viewGroup) {
            final Budget budget = mBudgetList.get(position);
            BudgetView budgetView = BudgetView.createBudgetView(mInflater, (BudgetView) view, viewGroup, budget);

            budgetView.setOnViewTransactionsClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    BudgetTransactionsFragment.newInstance(budget.id).show(getFragmentManager(), BudgetTransactionsFragment.TAG);
                }
            });
            budgetView.setOnEditClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    mCallback.onBudgetSelected(budget.id);
                }
            });
            budgetView.setOnBudgetClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    EditTransactionFragment.newInstance(budget.id).show(getFragmentManager(), EditTransactionFragment.TAG);
                }
            });

            return budgetView;
        }
    }
}

File: mobile/src/main/kotlin/com/bclymer/dailybudget/fragments/BudgetsFragment.kt
package com.bclymer.dailybudget.fragments

import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.GridView
import android.widget.TextView
import butterknife.bindView
import com.bclymer.dailybudget.R
import com.bclymer.dailybudget.database.BudgetRepository
import com.bclymer.dailybudget.events.BudgetUpdatedEvent
import com.bclymer.dailybudget.models.Budget
import com.bclymer.dailybudget.utilities.Util
import com.bclymer.dailybudget.views.BudgetView
import com.nhaarman.listviewanimations.appearance.AnimationAdapter
import com.nhaarman.listviewanimations.appearance.simple.ScaleInAnimationAdapter
import rx.android.schedulers.AndroidSchedulers
import java.util.*

/**
 * Created by bclymer on 9/26/2014.
 */
class BudgetsFragment() : BaseFragment(R.layout.fragment_budgets) {

    private val mGridView: GridView by bindView(android.R.id.list)
    private val mEmptyView: ViewGroup by bindView(android.R.id.empty)
    private val mButtonNewBudget: Button by bindView(R.id.fragment_budgets_emptyview_button_new_budget)
    private val mTextViewTotalBudget: TextView by bindView(R.id.fragment_budgets_textview_total)

    private var mAdapter: AnimationAdapter = ScaleInAnimationAdapter(BudgetAdapter())
    private var mBudgetList: MutableList<Budget> = arrayListOf()

    private val mCallback: BudgetSelectedCallback by lazy { activity as BudgetSelectedCallback }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        if (activity !is BudgetSelectedCallback) {
            throw RuntimeException("Activity $activity must implement BudgetSelectedCallback to display BudgetsFragment")
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        BudgetRepository.updateAndGetBudgets()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    mBudgetList = it.toMutableList()
                    updateTotalBudget()
                    mAdapter.notifyDataSetChanged()
                }

        mAdapter.setAbsListView(mGridView)
        mGridView.adapter = mAdapter
        mGridView.emptyView = mEmptyView
        mEventBus.register(this)

        mGridView.setOnItemClickListener { adapterView, view, position, l ->
            mCallback.onBudgetSelected(mBudgetList[position].id)
        }

        mButtonNewBudget.setOnClickListener {
            mCallback.onBudgetSelected(EditBudgetFragment.NO_BUDGET_ID_VALUE)
        }
    }

    override fun onDestroyView() {
        mEventBus.unregister(this)
        super.onDestroyView()
    }

    fun onEventMainThread(event: BudgetUpdatedEvent) {
        val copiedList = ArrayList(mBudgetList)
        for (budget in copiedList) {
            if (budget.id == event.budget.id) {
                if (event.deleted) {
                    mBudgetList.remove(event.budget)
                } else {
                    BudgetRepository.cloneBudget(from = event.budget, to = budget)
                }
                updateTotalBudget()
                mAdapter.notifyDataSetChanged()
                return
            }
        }
        mBudgetList.add(event.budget)
        updateTotalBudget()
        mAdapter.notifyDataSetChanged()
    }

    private fun updateTotalBudget() {
        mTextViewTotalBudget.text = "Total Budget: ${Util.makeLikeMoney(mBudgetList.sumByDouble { it.amountPerPeriod })}"
    }

    interface BudgetSelectedCallback {
        fun onBudgetSelected(budgetId: Int)
    }

    private inner class BudgetAdapter : BaseAdapter() {

        private val mInflater: LayoutInflater by lazy { activity.layoutInflater }

        override fun getCount() = mBudgetList.size

        override fun getItem(position: Int) = mBudgetList[position]

        override fun getItemId(position: Int) = position.toLong()

        override fun getView(position: Int, view: View?, viewGroup: ViewGroup): View {
            val budget = mBudgetList[position]
            val budgetView = BudgetView.createBudgetView(mInflater, view as? BudgetView, viewGroup, budget)

            budgetView.setOnViewTransactionsClickListener(View.OnClickListener { BudgetTransactionsFragment.newInstance(budget.id).show(fragmentManager, BudgetTransactionsFragment.TAG) })
            budgetView.setOnEditClickListener(View.OnClickListener { mCallback.onBudgetSelected(budget.id) })
            budgetView.setOnBudgetClickListener(View.OnClickListener { EditTransactionFragment.newInstance(budget.id).show(fragmentManager, EditTransactionFragment.TAG) })

            return budgetView
        }
    }

    companion object {

        fun newInstance(): BudgetsFragment {
            return BudgetsFragment()
        }
    }
}
